frostsnap_core/src/device.rs

```diff
@@ -206,44 +311,72 @@ impl FrostSigner {
-                    session_hash,
+                    agg_input,
+                    key_name: key_name.clone(),
                 });
 
-                Ok(vec![DeviceSend::ToUser(DeviceToUserMessage::CheckKeyGen {
-                    key_id: secret_share.public_key().key_id(),
-                    session_hash,
-                    key_name,
-                })])
+                Ok(vec![DeviceSend::ToUser(Box::new(
+                    DeviceToUserMessage::CheckKeyGen {
+                        key_id,
+                        session_hash,
+                        key_name,
+                    },
+                ))])
             }
             (None, CoordinatorToDeviceMessage::RequestSign(sign_req)) => {
-                let key_id = sign_req.key_id;
+                let root_key = sign_req.root_key;
+                let key_id = KeyId::from_root_key(root_key);
+                let access_structure_id = sign_req.access_structure_id;
                 let nonces = sign_req.nonces.clone();
-                let key = self.secret_shares.get(&key_id).ok_or_else(|| {
-                    Error::signer_invalid_message(
-                        &message,
-                        // we could instead send back a message saying we don't have this key but I
-                        // think this will never happen in practice unless we have a way for one
-                        // coordinator to delete a key from a device without the other coordinator
-                        // knowing.
-                        format!("device doesn't have key for {key_id}"),
-                    )
-                })?;
+                let key_data = self
+                    .keys
+                    .get(&key_id)
+                    .ok_or_else(|| {
+                        Error::signer_invalid_message(
+                            &message,
+                            // we could instead send back a message saying we don't have this key but I
+                            // think this will never happen in practice unless we have a way for one
+                            // coordinator to delete a key from a device without the other coordinator
+                            // knowing.
+                            format!("device doesn't have key for {key_id}"),
+                        )
+                    })?
+                    .clone();
+
+                let appkey = Appkey::derive_from_root_key(root_key);
+                let access_structure_data =
+                    key_data.access_structures.get(&access_structure_id)
+                        .ok_or_else( || {
+                            Error::signer_invalid_message(
+                                &message,
+                                format!("this device is not part of that access structure: {access_structure_id}"),
+                            )
+                        })?.clone();
+
+                // XXX We only support signing one share at a time for now
```

Comment:
This commit adds multiple shares per access structure. We'll need a refactor to be able to sign with multiple shares.